community-operators/spark-gcp/sparkoperator.2.4.0.clusterserviceversion.yaml

```diff
@@ -0,0 +1,299 @@
+      version: v1beta1
+      displayName: ScheduledSparkApplication
+      description: A configuration file for a ScheduledSparkApplication custom resource.
+    - kind: SparkApplication
+      name: sparkapplications.sparkoperator.k8s.io
+      version: v1beta1
+      displayName: SparkApplication
+      description: A configuration file for a SparkApplication custom resource.
+      resources:
+        - kind: Service
+          version: v1
+        - kind: Pod
+          version: v1
+      statusDescriptors:
+        - description: Current state of the application.
+          displayName: Application State
+          path: applicationState.state
+          x-descriptors: 
+            - 'urn:alm:descriptor:text'
+        - description: Indicates the error message in case of failures.
+          displayName: Error Message
+          path: applicationState.errorMessage
+          x-descriptors: 
+            - 'urn:alm:descriptor:io.kubernetes.phase:reason'
+  installModes:
+  - type: OwnNamespace
+    supported: true
+  - type: SingleNamespace
+    supported: true
+  - type: MultiNamespace
+    supported: false
+  - type: AllNamespaces
+    supported: false
+  install:
+    spec:
+      deployments:
+      - name: spark-operator
+        spec:
+          replicas: 1
+          selector:
+            matchLabels:
+              app.kubernetes.io/name: spark-operator
+              app.kubernetes.io/version: v2.4.0-v1beta1
+          strategy:
+            type: Recreate
+          template:
+            metadata:
+              creationTimestamp: null
+              initializers:
+                pending: []
+              labels:
+                app.kubernetes.io/name: spark-operator
+                app.kubernetes.io/version: v2.4.0-v1beta1
+            spec:
+              containers:
+              - args:
+                - -logtostderr
+                - -install-crds=false
+                - -namespace=$(WATCH_NAMESPACE)
+                image: gcr.io/spark-operator/spark-operator:v2.4.0-v1beta1-latest
+                imagePullPolicy: Always
+                name: spark-operator
+                resources: {}
+                env:
+                - name: WATCH_NAMESPACE
+                  valueFrom:
+                    fieldRef:
+                      fieldPath: metadata.annotations['olm.targetNamespaces']
+              serviceAccountName: spark-operator
+      permissions:
+      - serviceAccountName: spark
+        rules:
+          - apiGroups:
+            - ""
+            resources:
+            - "pods"
+            verbs:
+            - "*"
+          - apiGroups:
+            - ""
+            resources:
+            - "services"
+            verbs:
+            - "*"
+          - apiGroups:
+            - security.openshift.io
+            resources:
+            - securitycontextconstraints
+            verbs:
+            - use
+            resourceNames:
+            - anyuid
+      clusterpermissions:
```

Comment:
Does this operator still need cluster permissions if it's installed to a single namespace?

Reply:
I think yes because the spark operator still needs permissions to get nodes. And permissions for cluster-scoped resources like `nodes` can only be given via a ClusterRole:
```YAML
- apiGroups:
  - ""
  resources:
  - nodes
  verbs:
  - get
```

To be more fine grained you could reference a ClusterRole(that gives permissions for `nodes`) with just a RoleBinding to limit the scope for other namespaced resources to only the operator's namespace.
https://kubernetes.io/docs/reference/access-authn-authz/rbac/#rolebinding-and-clusterrolebinding
But I'm not sure how to ask for that in the CSV.

Reply:
As far as I know there isn't a way to do that in a CSV. In this case, having a cluster role for accessing nodes is fine, along with other cluster-scoped resources. However, "pods" falls under both permissions and clusterPermissions; is there any way clusterPermissions on pods could be reduced?

Reply:
Sorry for the late response. 
The `clusterPermissions` are for the service account of the spark-operator since the operator still needs permissions to CRUD the driver/executor pods for an application.
The `permissions` are for the service account for the driver/executor pods.

Perhaps I could bind another Role to the `spark-operator` service account for non cluster scoped resources like pods. And only keep the ClusterRole for nodes. 
I'll try it out. I'm not sure but I think I can request another Role/Rolebinding in `permissions` for the same `spark-operator` service account.